src/Cardano/Wallet/Network/HttpBridge.hs

```diff
@@ -132,79 +132,72 @@ fetchBlocksFromTip network start tipHash =
         else
             pure [block]
 
-
 {-------------------------------------------------------------------------------
                             HTTP-Bridge Client
 -------------------------------------------------------------------------------}
 
 -- | Endpoints of the cardano-http-bridge API.
-data HttpBridge m e = HttpBridge
+data HttpBridge m = HttpBridge
     { getBlock
-        :: Hash "BlockHeader" -> ExceptT e m Block
+        :: Hash "BlockHeader" -> ExceptT ErrNetworkUnreachable m Block
     , getEpoch
-        :: Word64 -> ExceptT e m [Block]
+        :: Word64 -> ExceptT ErrNetworkUnreachable m [Block]
     , getNetworkTip
-        :: ExceptT e m (Hash "BlockHeader", BlockHeader)
+        :: ExceptT ErrNetworkUnreachable m (Hash "BlockHeader", BlockHeader)
     , postSignedTx
-        :: SignedTx -> ExceptT e m ()
+        :: SignedTx -> ExceptT ErrPostTx m ()
     }
 
 -- | Construct a new network layer
 mkHttpBridge
-    :: Manager -> BaseUrl -> NetworkName -> HttpBridge IO HttpBridgeError
+    :: Manager -> BaseUrl -> NetworkName -> HttpBridge IO
 mkHttpBridge mgr baseUrl network = HttpBridge
-    { getBlock = \hash -> do
+    { getBlock = \hash -> ExceptT $ do
         hash' <- hashToApi' hash
-        run (getApiT <$> cGetBlock network hash')
-    , getEpoch = \ep ->
-        run (map getApiT <$> cGetEpoch network (EpochIndex ep))
-    , getNetworkTip =
-        run (blockHeaderHash <$> cGetNetworkTip network)
-    , postSignedTx =
-        -- TODO: We would be good to parse errors from the node
-        -- like "Failed to send to peers: Blockchain protocol error"
-        -- and "Transaction failed verification: transaction has no inputs"
-        void . run . (cPostSignedTx network) . ApiT
+        run (getApiT <$> cGetBlock network hash') >>= defaultHandler
+
+    , getEpoch = \ep -> ExceptT $ do
+        run (map getApiT <$> cGetEpoch network (EpochIndex ep)) >>= \case
+            Left (FailureResponse e) | responseStatusCode e == status404 ->
+                return $ Right []
+            x -> defaultHandler x
+
+    , getNetworkTip = ExceptT $ do
+        run (blockHeaderHash <$> cGetNetworkTip network) >>= defaultHandler
+
+    , postSignedTx = \tx -> void $ ExceptT $ do
+        let e0 = "Failed to send to peers: Blockchain protocol error"
+        run (cPostSignedTx network (ApiT tx)) >>= \case
+            Left (FailureResponse e)
+                | responseStatusCode e == status400 -> do
+                    let msg = T.decodeUtf8 $ BL.toStrict $ responseBody e
+                    return $ Left $ ErrPostTxBadRequest msg
+            Left (FailureResponse e)
+                | responseStatusCode e == status500 && responseBody e == e0 -> do
+                    let msg = T.decodeUtf8 $ BL.toStrict $ responseBody e
+                    return $ Left $ ErrPostTxProtocolFailure msg
+            x -> left ErrPostTxNetworkUnreachable <$> defaultHandler x
```

Comment:
Is this safe? Thinking if `e0` is changed on the node-side, or some other error is added? `ErrPostTxNetworkUnreachable` seems to capture the the message at least so should be fine for now.

Reply:
> Thinking if e0 is changed on the node-side

We do fix the cardano-http-bridge version as part of our deployment. So this only holds for version we have tested against. And, in this particular version, it cannot change. If it change later, then we'll have our test catching it, hopefully.